### PR TITLE
cache metal compute pipelines

### PR DESCRIPTION
**Pull request by mikewoodworth**
instead of recreating in each invocation of halide_metal_run

"Creating compute pipeline state objects can be expensive, so you usually create these pipeline objects when your app is initialized and reuse them throughout its lifetime."
- https://developer.apple.com/documentation/metal/mtlcomputepipelinestate?language=objc

This pull also moves the call to halide_start_clock out of halide_metal_acquire_context so that clients that overload that method don't experience crashes when running in debug.

### CODE REVIEW

File: src/runtime/metal.cpp [hunk 1]
```diff
@@ -714,18 +734,42 @@ WEAK int halide_metal_run(void *user_context,
     halide_assert(user_context, state_ptr);
     module_state *state = (module_state*)state_ptr;
 
-    mtl_function *function = new_function_with_name(state->library, entry_name, strlen(entry_name));
-    if (function == 0) {
-        error(user_context) << "Metal: Could not get function " << entry_name << "from Metal library.\n";
-        return -1;
+    
+    cached_pipeline_state *cached_pipeline = state->cached_pipeline_state_list;
+    while (cached_pipeline != NULL && strcmp(cached_pipeline->function_name, entry_name) != 0) {
```
Comment (shoaibkamil) [inline src/runtime/metal.cpp]: I'd change this to `strncmp`.

File: src/runtime/metal.cpp [hunk 2]
```diff
@@ -714,18 +734,42 @@ WEAK int halide_metal_run(void *user_context,
...
+        }
+ 
+        cached_pipeline_state *new_pipeline = (cached_pipeline_state*)malloc(sizeof(cached_pipeline_state));
+        new_pipeline->function_name = entry_name;
+        new_pipeline->precompiled_pipeline_state = pipeline_state;
+        new_pipeline->next = state->cached_pipeline_state_list;
+        state->cached_pipeline_state_list = new_pipeline;
+        
+        cached_pipeline = new_pipeline;
+    }else{
```
Comment (shoaibkamil) [inline src/runtime/metal.cpp]: Spacing

File: src/runtime/metal.cpp [hunk 3]
```diff
@@ -714,18 +734,42 @@ WEAK int halide_metal_run(void *user_context,
...
+    while (cached_pipeline != NULL && strcmp(cached_pipeline->function_name, entry_name) != 0) {
+        cached_pipeline = cached_pipeline->next;
     }
 
-    mtl_compute_pipeline_state *pipeline_state = new_compute_pipeline_state_with_function(metal_context.device, function);
-    if (pipeline_state == 0) {
-        error(user_context) << "Metal: Could not allocate pipeline state.\n";
-        return -1;
+    
+    if(!cached_pipeline){
```
Comment (shoaibkamil) [inline src/runtime/metal.cpp]: Why not just create the pipeline states during `halide_metal_initialize_kernel()`?

Comment (shoaibkamil) [reply src/runtime/metal.cpp]: Also, spacing.

**Comment by shoaibkamil** [comment]
In principle, I agree with this change-- I had a PR that did this (#2580), but at the time we didn't merge it in, so I defer to @zvookin.

**Comment by mikewoodworth** [comment]
> In principle, I agree with this change-- I had a PR that did this (#2580), but at the time we didn't merge it in, so I defer to @zvookin.

I'll wait to see what @zvookin says before making changes.

**Comment by zvookin** [comment]
Will review shortly. I need to go page back in some knowledge on this. I recall a PR for OpenCL recently where there was some debate about what is supposed to be cheap and what is supposed to be expensive and what state has to be recreated under certain circumstances. E.g. is there any situation were we have to recreate the pipeline state for the kernel invocation to be valid? (E.g. if the machine switches from discrete GPU to integrated between two calls? Usually Apple's stuff takes care of that internally, but I have to go check the docs.)

Ultimately I'd like to surface this stuff via an explicit (and general) API and allow end user control, but it is clear we will need some changes like this for a while.

**Comment by mikewoodworth** [comment]
In the default implementation halide_default_metal_acquire_context will return the preexisting device for all later invocations... we should be on the same device for all subsequent calls to acquire. But I suppose that problem could arise when overloading halide_metal_acquire_context.

We can cache device and reuse for later invocations, or do a simple match (cached == acquired) and if not recompile. 

If we're worrying about device changing mid use, I suspect there are other concerns with buffers being on the wrong device, but I'd have to examine the source more closely.

**Comment by steven-johnson** [comment]
Ready for final review?

**Comment by mikewoodworth** [comment]
I haven't made any of @shoaibkamil fixes yet, waiting to hear if @zvookin is ok with the general direction first.

**Comment by steven-johnson** [comment]
Ping to see if still waiting on feedback.

**Comment by steven-johnson** [comment]
What's the status on this PR -- still waiting on feedback?

**Comment by mikewoodworth** [comment]
confirm with @zvookin, but I think its a non-starter

**Pull request closed without merging**